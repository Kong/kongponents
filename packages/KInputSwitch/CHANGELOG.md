# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.1.1-2](https://github.com/Kong/kongponents/compare/@kongponents/kinputswitch@0.1.1-1...@kongponents/kinputswitch@0.1.1-2) (2019-11-08)

**Note:** Version bump only for package @kongponents/kinputswitch





# @kongponents/kinputswitch

## Table of Contents

- [0.1.1](#011)

### 0.1.1
:rocket: initial release [#106](https://github.com/Kong/kongponents/pull/106)
