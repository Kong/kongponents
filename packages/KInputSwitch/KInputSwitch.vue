<template>
  <label class="k-switch">
    <input
      :checked="value"
      type="checkbox"
      v-on="listeners">
    <div class="switch-control"/>
    <span v-if="label || $scopedSlots.label">
      <slot name="label">{{ label }}</slot>
    </span>
  </label>
</template>

<script>
export default {
  name: 'KInputSwitch',
  props: {
    /**
     * Sets whether or not toggle is checked
     */
    value: {
      type: Boolean,
      default: false,
      required: true
    },

    /**
     * Overrides default on/off label text
     */
    label: {
      type: String,
      default: ''
    }
  },

  computed: {
    toggleText () {
      return this.value
        ? 'on'
        : 'off'
    },

    listeners () {
      const listeners = {
        ...this.$listeners,
        change: e => {
          this.$emit('change', e.target.checked)
          this.$emit('input', e.target.checked)
        }
      }

      // remove 'input' since it doesn't need to be bound to the checkbox
      delete listeners['input']

      return listeners
    }
  }
}
</script>

<style lang="scss" scoped>
@import '~@kongponents/styles/_variables.scss';
@import '~@kongponents/styles/forms/_switch.scss';
</style>
