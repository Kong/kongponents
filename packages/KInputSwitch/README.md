# @kongponents/kinputswitch

[![](https://img.shields.io/npm/v/@kongponents/kinputswitch.svg?style=flat-square)](https://www.npmjs.com/package/@kongponents/kinputswitch)

```vue
<KInputSwitch 
  v-model="checked" 
  :label="checked ? 'on' : 'off'" />
```
