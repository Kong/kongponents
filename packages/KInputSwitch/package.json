{
  "name": "@kongponents/kinputswitch",
  "version": "0.2.1",
  "description": "KInputSwitch description here.",
  "main": "dist/KInputSwitch.umd.min.js",
  "componentName": "KInputSwitch",
  "source": "KInputSwitch.vue",
  "files": [
    "dist",
    "KInputSwitch.vue"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme"
}
