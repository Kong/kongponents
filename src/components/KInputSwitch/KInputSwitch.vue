<template>
  <KTooltip
    v-if="disabled && disabledTooltipText"
    :label="disabledTooltipText"
  >
    <label
      class="k-switch k-input-switch"
      :class="[$attrs.class]"
      :disabled="disabled"
      :for="$attrs.id ? String($attrs.id) : undefined"
    >
      <span v-if="(label || $slots.label) && labelPosition === 'left'">
        <slot name="label">{{ label }}</slot>
      </span>

      <input
        :checked="modelValue"
        :disabled="disabled"
        type="checkbox"
        v-bind="strippedAttrs"
        @change="handleChange"
        @input="handleChange"
      >
      <div :class="['switch-control', labelPosition === 'right' ? 'has-label-right' : 'has-label-left' ]" />

      <span v-if="(label || $slots.label) && labelPosition === 'right'">
        <slot name="label">{{ label }}</slot>
      </span>
    </label>
  </KTooltip>

  <label
    v-else
    class="k-switch k-input-switch"
    :class="[$attrs.class, { 'switch-with-icon' : enabledIcon }]"
    :disabled="disabled ? disabled : undefined"
    :for="$attrs.id ? String($attrs.id) : undefined"
  >
    <span v-if="(label || $slots.label) && labelPosition === 'left'">
      <slot name="label">{{ label }}</slot>
    </span>

    <input
      :checked="modelValue"
      :disabled="disabled ? disabled : undefined"
      v-bind="strippedAttrs"
      type="checkbox"
      @change="handleChange"
      @input="handleChange"
    >

    <div :class="['switch-control', labelPosition === 'right' ? 'has-label-right' : 'has-label-left' ]" />

    <KIcon
      v-if="enabledIcon && modelValue === true"
      :color="`var(--white, var(--kui-color-text-inverse, ${KUI_COLOR_TEXT_INVERSE}))`"
      icon="check"
    />

    <span v-if="(label || $slots.label) && labelPosition === 'right'">
      <slot name="label">{{ label }}</slot>
    </span>
  </label>
</template>

<script lang="ts" setup>
import type { PropType } from 'vue'
import { computed, useAttrs } from 'vue'
import KTooltip from '@/components/KTooltip/KTooltip.vue'
import KIcon from '@/components/KIcon/KIcon.vue'
import type { LabelPosition } from '@/types'
import { LabelPositionArray } from '@/types'
import { KUI_COLOR_TEXT_INVERSE } from '@kong/design-tokens'

const props = defineProps({
  /**
  * Sets whether or not toggle is checked
  */
  modelValue: {
    type: Boolean,
    default: false,
    required: true,
  },
  /**
  * Overrides default on/off label text
  */
  label: {
    type: String,
    default: '',
  },
  /**
  * Should the switch be positioned to the left or right of the label
  */
  labelPosition: {
    type: String as PropType<LabelPosition>,
    default: 'right',
    validator: (position: LabelPosition): boolean => LabelPositionArray.includes(position),
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  /**
  * Tooltip text to be displayed if the switch is disabled
  */
  disabledTooltipText: {
    type: String,
    default: '',
  },
  /**
  * Sets whether or not to display a check icon if the switch is enabled
  */
  enabledIcon: {
    type: Boolean,
    default: false,
  },
})

const emit = defineEmits<{
  (e: 'change', val: boolean): void
  (e: 'input', val: boolean): void
  (e: 'update:modelValue', val: boolean): void
}>()

const attrs = useAttrs()

/**
* Strips falsy `disabled` attribute, so it does not fall onto native <a> elements.
* Vue 3 no longer removes attribute if the value is boolean false. Instead, it's set as attr="false".
* So for <KButton :disabled="false" to="SOME_URL">, the rendered <a> element will have `disabled="false"`,
* which is greyed out and cannot be interacted with.
*/
const strippedAttrs = computed((): typeof attrs => {
  if (props.disabled !== undefined && props.disabled !== false) {
    return attrs
  }

  const modifiedAttrs = Object.assign({}, attrs)

  delete modifiedAttrs.class
  delete modifiedAttrs.disabled

  return modifiedAttrs
})

const handleChange = (event: Event): void => {
  if (props.modelValue !== (event.target as HTMLInputElement).checked) {
    emit('change', (event.target as HTMLInputElement).checked)
    emit('input', (event.target as HTMLInputElement).checked)
    emit('update:modelValue', (event.target as HTMLInputElement).checked)
  }
}
</script>

<script lang="ts">
export default {
  inheritAttrs: false,
}
</script>

<style lang="scss" scoped>
@import '@/styles/variables';
@import '@/styles/tmp-variables';

$kInputSwitchWidth: 44px;
$kInputSwitchHeight: 24px;
$kInputSwitchTransition: $tmp-animation-timing-2 linear;

.k-switch {
  align-items: center;
  cursor: pointer;
  display: inline-flex;
  position: relative;

  .has-label-left + .kong-icon {
    margin-left: var(--kui-space-40, $kui-space-40);
  }

  .has-label-right + .kong-icon {
    left: 56px;
    position: absolute;
  }

  .kong-icon {
    top: 0;
    transform: translateX(-54px);
  }

  &.switch-with-icon .switch-control {
    width: 48px;
  }

  &.switch-with-icon .kong-icon {
    height: 20px;
    left: 57px;
    width: 22px;
  }

  &.switch-with-icon input:checked + .switch-control:after {
    left: 26px;
  }

  .switch-control {
    background-color: var(--KInputSwitchBackground, var(--grey-400, $kui-color-background-neutral-weak));
    border-radius: 12px;
    display: block;
    height: $kInputSwitchHeight;
    margin-right: var(--kui-space-60, $kui-space-60);
    position: relative;
    transition: $kInputSwitchTransition;
    width: $kInputSwitchWidth;

    &.has-label-left {
      margin-left: var(--kui-space-60, $kui-space-60);
      margin-right: var(--kui-space-0, $kui-space-0);
    }

    // Toggle
    &:after {
      background-color: var(--kui-color-background, $kui-color-background);
      border-radius: var(--kui-border-radius-circle, $kui-border-radius-circle);
      content: "";
      display: block;
      height: 20px;
      left: 2px;
      position: absolute;
      top: 2px;
      transition: $kInputSwitchTransition;
      width: 20px;
    }
  }

  &[disabled]:not(:disabled) {
    cursor: not-allowed;
    .switch-control,
    input {
      opacity: 0.3;
      pointer-events: none;
    }
  }

  // Hide default checkbox
  input {
    display: none;
    &:checked + .switch-control {
      background-color: $tmp-color-green-500;
      &:after {
        left: calc($kInputSwitchWidth / 2);
      }
    }
  }

  span {
    color: var(--kui-color-text, $kui-color-text);
  }
}
</style>
