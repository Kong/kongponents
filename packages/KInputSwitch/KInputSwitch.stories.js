import { storiesOf } from '../../storybook/node_modules/@storybook/vue/dist/client'
import KInputSwitch from './KInputSwitch.vue'

storiesOf('KInputSwitch', module)
  .add('Default', () => ({
    components: { KInputSwitch },
    template: `
      <KInputSwitch
        v-model="checked"
        @change="handleToggle"/>
    `,
    data: () => ({
      checked: false
    }),
    methods: {
      handleToggle (value) {
        this.checked = value
        console.log(this.checked)
      }
    }
  }), { info: {} })
