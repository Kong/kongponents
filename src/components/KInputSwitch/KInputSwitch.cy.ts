import KInputSwitch from '@/components/KInputSwitch/KInputSwitch.vue'

describe('KInputSwitch', () => {
  it('shows as checked when prop passed', () => {
    cy.mount(KInputSwitch, {
      props: {
        modelValue: true,
      },
    })

    cy.get('input').should('be.checked')
  })

  it('emits checked value on click', () => {
    cy.mount(KInputSwitch, {
      props: {
        modelValue: true,
      },
    })

    cy.getTestId('switch-control').click().then(() => {
      cy.wrap(Cypress.vueWrapper.emitted()).should('have.property', 'change')
      cy.wrap(Cypress.vueWrapper.emitted()).should('have.property', 'input')
      cy.wrap(Cypress.vueWrapper.emitted()).should('have.property', 'update:modelValue')
    })
  })

  it('does not emit checked value on click when disabled', () => {
    cy.mount(KInputSwitch, {
      props: {
        modelValue: true,
        disabled: true,
      },
    })

    cy.getTestId('switch-control').click().then(() => {
      cy.wrap(Cypress.vueWrapper.emitted()).should('not.have.property', 'change')
      cy.wrap(Cypress.vueWrapper.emitted()).should('not.have.property', 'input')
      cy.wrap(Cypress.vueWrapper.emitted()).should('not.have.property', 'update:modelValue')
    })
  })
})
