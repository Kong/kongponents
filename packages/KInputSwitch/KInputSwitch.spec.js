import { mount } from '@vue/test-utils'
import KInputSwitch from '@/KInputSwitch/KInputSwitch'

describe('KInputSwitch', () => {
  it('shows as checked when prop passed', () => {
    const wrapper = mount(KInputSwitch, {
      propsData: {
        isChecked: true
      }
    })

    expect(wrapper.find('input').element.checked).toBe(true)
  })

  it('emits checked value on click', () => {
    const wrapper = mount(KInputSwitch, {
      propsData: {
        isChecked: true
      }
    })
    const toggle = wrapper.find('input')

    toggle.trigger('click')

    expect(wrapper.emitted().change).toBeTruthy()
    expect(wrapper.emitted().change).toHaveLength(1)
  })

  it('matches snapshot', () => {
    const wrapper = mount(KInputSwitch, {
      propsData: {
        isChecked: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
