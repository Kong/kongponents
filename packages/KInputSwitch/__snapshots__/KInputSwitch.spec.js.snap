// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KInputSwitch matches snapshot 1`] = `
<label class="k-switch"><input type="checkbox">
  <div class="switch-control"></div>
  <!---->
  <!---->
</label>
`;
